go-controller/observability-lib/parse_sample.go

```diff
@@ -0,0 +1,241 @@
+		writer = bufio.NewWriter(file)
+	} else {
+		writer = os.Stdout
+	}
+	l := log.New(writer, "", log.Ldate|log.Ltime|log.Lmicroseconds)
+	printlnFunc := func(a ...any) {
+		l.Println(a...)
+	}
+
+	fam, err := netlink.GenlFamilyGet(PSAMPLE_GENL_NAME)
+	if err != nil {
+		return fmt.Errorf("error getting netlink family %s: %w", PSAMPLE_GENL_NAME, err)
+	}
+	if len(fam.Groups) == 0 {
+		return fmt.Errorf("no mcast groups found for %s", PSAMPLE_GENL_NAME)
+	}
+	var ovsGroupID uint32
+	for _, group := range fam.Groups {
+		if group.Name == PSAMPLE_NL_MCGRP_SAMPLE_NAME {
+			ovsGroupID = group.ID
+		}
+	}
+	if ovsGroupID == 0 {
+		return fmt.Errorf("no mcast group found for %s", PSAMPLE_NL_MCGRP_SAMPLE_NAME)
+	} else {
+		fmt.Printf("Found group %s, id %d\n", PSAMPLE_NL_MCGRP_SAMPLE_NAME, ovsGroupID)
+	}
+	sock, err := nl.Subscribe(nl.GENL_ID_CTRL, uint(ovsGroupID))
+	if err != nil {
+		return fmt.Errorf("error subscribing to netlink group %d: %w", ovsGroupID, err)
+	}
+
+	// Otherwise sock.Receive() will be blocking and won't return on context close
+	if err = unix.SetNonblock(sock.GetFd(), true); err != nil {
+		return fmt.Errorf("error setting non-blocking mode: %w", err)
+	}
+
+	defer func() {
+		sock.Close()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			msgs, _, err := sock.Receive()
+			if err != nil {
+				if err == syscall.EAGAIN {
+					continue
+				}
+				printlnFunc("ERROR: receive failed:", err)
+				continue
+			}
+			if err = r.parseMsg(msgs, printlnFunc); err != nil {
+				printlnFunc("ERROR: ", err)
+			}
+		}
+	}
+}
+
+func getHostEndian() binary.ByteOrder {
+	buf := [2]byte{}
+	*(*uint16)(unsafe.Pointer(&buf[0])) = uint16(0xABCD)
+
+	switch buf {
+	case [2]byte{0xCD, 0xAB}:
+		return binary.LittleEndian
+	case [2]byte{0xAB, 0xCD}:
+		return binary.BigEndian
+	default:
+		panic("Could not determine native endianness.")
+	}
+}
+
+var hostEndian = getHostEndian()
+
+func (r *SampleReader) parseMsg(msgs []syscall.NetlinkMessage, printlnFunc func(a ...any)) error {
+	var packetStr, sampleStr string
+	for _, msg := range msgs {
+		data := msg.Data[nl.SizeofGenlmsg:]
+		for attr := range nl.ParseAttributes(data) {
+			if r.logCookie && attr.Type == PSAMPLE_ATTR_SAMPLE_GROUP {
+				if uint64(len(attr.Value)) == 4 {
+					g := uint32(0)
+					// group is encoded using host endian
+					err := binary.Read(bytes.NewReader(attr.Value), hostEndian, &g)
+					if err != nil {
+						return err
+					}
+					r.cookieStr[0] = fmt.Sprintf("group_id=%v", g)
+				}
+			}
+			if r.logCookie || r.decoder != nil {
+				if attr.Type == PSAMPLE_ATTR_USER_COOKIE {
+					if uint64(len(attr.Value)) == sampledecoder.CookieSize {
+						c := sampledecoder.Cookie{}
+						err := binary.Read(bytes.NewReader(attr.Value), sampledecoder.SampleEndian, &c)
+						if err != nil {
+							return err
+						}
+						if r.logCookie {
+							r.cookieStr[1] = fmt.Sprintf("obs_domain=%v, obs_point=%v", c.ObsDomainID, c.ObsPointID)
```

Comment:
Nit: Maybe it's because I'm used to 80 column line lengths but I think I'd add `parseSampleGroup()/parseUserCookie()/parseData()` functions.

Reply:
heheh yeah I see what you mean, but there are some message-scoped variables that are set inside potential parsing sub-functions, and I feel like it is easier to follow when it is in the same function body...
But! I can make the lines a bit shorter :D

Reply:
Ack, thanks.